orttraining/orttraining/training_ops/cuda/communication/send.cc

```diff
@@ -3,69 +3,130 @@
 
 namespace onnxruntime {
 namespace cuda {
 
-#define REGISTER_SEND_KERNEL_TYPED(T)                                    \
-  ONNX_OPERATOR_TYPED_KERNEL_EX(                                         \
-      Send,                                                              \
-      kMSDomain,                                                         \
-      1,                                                                 \
-      T,                                                                 \
-      kCudaExecutionProvider,                                            \
-      KernelDefBuilder()                                                 \
-          .InputMemoryType<OrtMemTypeCPUInput>(0)   /* CPU variable */   \
-          .OutputMemoryType<OrtMemTypeCPUOutput>(0) /* CPU variable */   \
-          .TypeConstraint("T", DataTypeImpl::GetTensorType<T>())         \
-          .TypeConstraint("TBool", DataTypeImpl::GetTensorType<bool>()), \
-      Send<T>);
-
-REGISTER_SEND_KERNEL_TYPED(MLFloat16)
-REGISTER_SEND_KERNEL_TYPED(float)
-REGISTER_SEND_KERNEL_TYPED(double)
+ONNX_OPERATOR_KERNEL_EX(
+    Send,
+    kMSDomain,
+    1,
+    kCudaExecutionProvider,
+    KernelDefBuilder()
+        .InputMemoryType<OrtMemTypeCPUInput>(0)   /* CPU variable */
+        .InputMemoryType<OrtMemTypeCPUInput>(1)   /* CPU variable */
+        .OutputMemoryType<OrtMemTypeCPUOutput>(0) /* CPU variable */
+        .TypeConstraint("TBool", DataTypeImpl::GetTensorType<bool>())
+        .TypeConstraint("TInt64", DataTypeImpl::GetTensorType<int64_t>())
+        .TypeConstraint("V", DataTypeImpl::AllFixedSizeTensorTypes()),
+    Send);
 
 void CUDART_CB HostSend(void* args) {
   CommInfo_t* info = reinterpret_cast<CommInfo_t*>(args);
-  int mpi_code  = MPI_Send(info->buffer, info->size, MPI_CHAR, info->rank, info->tag, MPI_COMM_WORLD);
+  int mpi_code = MPI_Send(info->buffer, info->size, MPI_CHAR, info->rank, info->tag, MPI_COMM_WORLD);
   ORT_ENFORCE(mpi_code == MPI_SUCCESS, "MPI Send fails.");
 }
 
-template <typename T>
-Status Send<T>::ComputeInternal(OpKernelContext* ctx) const {
+Status Send::ComputeInternal(OpKernelContext* ctx) const {
   // Check if control signal is true.
   const Tensor* input_signal_tensor = ctx->Input<Tensor>(0);
   const bool* input_signal = input_signal_tensor->template Data<bool>();
   ORT_ENFORCE(*input_signal, "Input control signal of Send must be true before executing the node.");
 
-  // Start the communication.
-  const Tensor* x_tensor = ctx->Input<Tensor>(1);
-  size_t tensor_shape_size = static_cast<size_t>(x_tensor->Shape().NumDimensions());;
-  size_t tensor_size = static_cast<size_t>(x_tensor->Shape().Size());
-
+  // Extract remote rank
+  const Tensor* remote_rank_tensor = ctx->Input<Tensor>(1);
+  const int64_t* remote_rank = remote_rank_tensor->template Data<int64_t>();
+  const int dst = static_cast<int>(*remote_rank);
+
+  // Create buffers
+  const int tensor_num = static_cast<int>(element_types_.size());
```

Comment:
>element_types_ [](start = 42, length = 14)

if element_types_ is only used for getting number of tensor, it seems unnecessary to have this attribute... 

Consider enforcing input tensor type == element_type

Reply:
It is also used for type inference in onnx.

Reply:
I just added type checking in onnx def and size checking during graph resolve, so we don't need to do that in runtime. 
It will be slightly faster.